identity.go

```diff
@@ -33,19 +33,19 @@ type provisionalIdentity struct {
 	PrivateEncryptionKey []byte `json:"private_encryption_key"`
 }
 
-func newIdentity(config config, userIDString string) (*identity, error) {
+func newIdentity(config config, userIDString string) (identity, error) {
 	generatedAppID := newAppId(config.AppSecret)
 
 	if !bytes.Equal(generatedAppID, config.AppID) {
-		return nil, errors.New("App secret and app ID mismatch")
+		return identity{}, errors.New("App secret and app ID mismatch")
```

Comment:
I'm thorned here. I find nil more explicit about the value being unusable. IMO `new*` is used to when a library returns an Interface and the caller doesn't need to known how to initialise the internals. What do you think about taking an uninitialised struct as argument and filling it
```go
func fillIdentity(config config, userIDString string, placeholder *identity) (error) {
	generatedAppID := newAppId(config.AppSecret)
	if !bytes.Equal(generatedAppID, config.AppID) {
		return errors.New("App secret and app ID mismatch")
	}
	userID := hashUserID(config.AppID, userIDString)
	userSecret := newUserSecret(userID)

	epubSignKey, eprivSignKey, err := ed25519.GenerateKey(nil)
	if err != nil {
		return err
	}
	// ...
	identity.publicIdentity = {
			TrustchainID: config.AppID,
			Target:       "user",
			Value:        base64.StdEncoding.EncodeToString(userID),
	}
	// ...
	return nil
}
```

Reply:
I have to say I don't quite agree here. Having a `new_` function precisely conveys that the zero-value of `identity` is not usable as is, and one should not initialize it another way. The return value of the function can be empty as long as the error is not nil when the value is not usable, and I don't think we should have a pointer return only to be able to return nil.
A `fillIdentity` function seems a bit off to me as it explodes the logic into two parts (the caller and the callee) where you should never have the first without the second. I'd leave it that way but we can discuss this if need be.

@maximerety opinion?

Reply:
I have to admit I'm not a big fan of the fill version either.

I'm fine with the current change proposed by @mlevieux as well as with the previous method of returning a pointer.

Both are easy to use, and anyway this is an internal method so we shouldn't be too confused by what we return for ourselves.

Reply:
> anyway this is an internal method so we shouldn't be too confused by what we return for ourselves

true, let's go for the current change